### PR TITLE
feat: read assets from podlets using 103 early hints

### PR DESCRIPTION
**Pull request by digitalsadhu**
Adds support for reading assets from podlets via 103 early hints. Does not change the API surface for developers, this is purely an under the hood change.

* Implements phase 1: https://github.com/podium-lib/client/pull/410#issuecomment-2323536976
* [x] Needs: https://github.com/podium-lib/utils/pull/260
* Replaces: https://github.com/podium-lib/client/pull/410

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -71,3 +73,45 @@ export const filterAssets = (scope, assets) => {
...
+        const [href, ...rest] = parts;
+        const value = href.replace(/<|>|"/g, '').trim();
+
+        const asset = { value };
+        for (const key of rest) {
+            const [keyName, keyValue] = key.split('=');
+            asset[keyName.trim()] = keyValue.trim();
+        }
+
+        if (asset['asset-type'] === 'script') {
```
Comment (digitalsadhu) [inline lib/utils.js:93]: detect asset type using fields provided by this PR https://github.com/podium-lib/utils/pull/260

File: lib/http-outgoing.js [hunk 2]
```diff
@@ -120,6 +122,8 @@ export default class PodletClientHttpOutgoing extends PassThrough {
         this.#manifest = {
             // @ts-expect-error Internal property
             _fallback: '',
+            _js: [],
+            _css: [],
```
Comment (digitalsadhu) [inline lib/http-outgoing.js:126]: Cache properties for js and css.
These get (re)populated any time the fallback route is fetched

File: lib/http-outgoing.js [hunk 3]
```diff
@@ -146,6 +150,24 @@ export default class PodletClientHttpOutgoing extends PassThrough {
...
+
+    get css() {
+        // return the internal css value or, fallback to the manifest for backwards compatibility
+        return this.#css || this.#manifest.css;
+    }
+
+    set css(value) {
+        this.#css = value;
+    }
+
```
Comment (digitalsadhu) [inline lib/http-outgoing.js:170]: These are js and css getter/setters that are used whenever the content route is fetched OR when pushFallback is called (at which point the cached js and css are set)

File: lib/http-outgoing.js [hunk 4]
```diff
@@ -308,6 +330,10 @@ export default class PodletClientHttpOutgoing extends PassThrough {
     pushFallback() {
         // @ts-expect-error Internal property
         this.push(this.#manifest._fallback);
+        // @ts-expect-error Internal property
+        this.js = this.#manifest._js;
+        // @ts-expect-error Internal property
+        this.css = this.#manifest._css;
```
Comment (digitalsadhu) [inline lib/http-outgoing.js:336]: We've fallen back so we grab the cached fallback js and css and set it on the outgoing.js and outgoing.css properties

File: lib/resolver.content.js [hunk 5]
```diff
@@ -140,6 +142,22 @@ export default class PodletClientContentResolver {
...
+                    );
+                    const styleObjects = parsedAssetObjects.filter(
+                        (asset) => asset instanceof AssetCss,
+                    );
+                    // set the content js asset objects
+                    outgoing.js = filterAssets('content', scriptObjects);
+                    // set the content css asset objects
+                    outgoing.css = filterAssets('content', styleObjects);
+                }
+            },
```
Comment (digitalsadhu) [inline lib/resolver.content.js:160]: this onInfo block is called when early hints are received from the podlet during a content route request. AssetJs and AssetCss objects are created from the link header and then set on outgoing.js and outgoing.css

File: lib/resolver.fallback.js [hunk 6]
```diff
@@ -100,6 +102,30 @@ export default class PodletClientFallbackResolver {
...
+                        scriptObjects,
+                    );
+                    // set the fallback css asset fallback objects
+                    // @ts-expect-error internal property
+                    outgoing.manifest._css = filterAssets(
+                        'fallback',
+                        styleObjects,
+                    );
+                }
+            },
```
Comment (digitalsadhu) [inline lib/resolver.fallback.js:128]: this onInfo block is called when the fallback is fetched and the podlet sends back asset early hints. The js and css received is cached for later use whenever a content route fetch fails.

File: lib/http-outgoing.js [hunk 7]
```diff
@@ -146,6 +150,24 @@ export default class PodletClientHttpOutgoing extends PassThrough {
         this.#redirectable = redirectable;
     }
 
+    get js() {
+        // return the internal js value or, fallback to the manifest for backwards compatibility
+        return this.#js || this.#manifest.js;
```
Comment (wkillerud) [inline lib/http-outgoing.js:155]: Should the manifest property be prefixed with underscore? I see it is in the fallback.

```suggestion
        return this.#js || this.#manifest._js;
```

File: lib/http-outgoing.js [hunk 8]
```diff
@@ -146,6 +150,24 @@ export default class PodletClientHttpOutgoing extends PassThrough {
...
+    }
+
+    set js(value) {
+        this.#js = value;
+    }
+
+    get css() {
+        // return the internal css value or, fallback to the manifest for backwards compatibility
+        return this.#css || this.#manifest.css;
```
Comment (wkillerud) [inline lib/http-outgoing.js:164]:
```suggestion
        return this.#css || this.#manifest._css;
```

File: lib/resolver.content.js (hunk 5 above)
Comment (trygve-lie) [reply lib/resolver.content.js:160]: I think we might need some kinda guard that only accepts one `onInfo` and `103` http event here.

File: lib/resolver.fallback.js (hunk 6 above)
Comment (trygve-lie) [reply lib/resolver.fallback.js:128]: Same as on the content; I think we might need some kinda guard that only accepts one onInfo and 103 http event here.

**Comment by trygve-lie** [comment]
Great work. Super nice that we're not touching the public API atm.

File: lib/http-outgoing.js (hunk 7 above)
Comment (digitalsadhu) [reply lib/http-outgoing.js:155]: No, what's happening here is that our new feature stores js on this.#js (both on content or fallback) and if present (ie, the podlet sent early hints), we return it. If it's not, we fallback to the old way of doing things which was grabbing the js value from the manifest file.

File: lib/http-outgoing.js (hunk 8 above)
Comment (digitalsadhu) [reply lib/http-outgoing.js:164]: No, same as for js

**Review by wkillerud — APPROVED** [review]

**Pull request merged**